### PR TITLE
Dynamic time warping - aligners and distances

### PR DESCRIPTION
**Pull request by fkiraly**
This PR implements the first two thirds of [STEP10 - alignment, time warping, and dtw based clustering/classification](https://github.com/sktime/enhancement-proposals/blob/main/steps/10_alignment_warping_knn/step.md).

It introduces:

* a new estimator scitype, the aligner, and a new object scitype, the `Alignment`. It produces "alignments" for sequences/series in a `Panel`. This is located in the new `alignment` module.
* Checkers for alignments and input validation in the `datatypes` module, for the new `Alignment` scitype. The two new mtypes correspond to alignments in `iloc` and `loc` coordinates.
* aligners which interface the `dtw-python` package (this is introduced as a *soft dependency*). There are two in the `dtw_python` sub-module, one which takes the default distance(s) and does univariate pairwise alignment, and one which can take any pairwise distance (for tabular/primitive data) and do pairwise alignment of multivariate series.
* a distance `DistFromAligner` for time series, which takes any aligner and produces a matrix of pairwise alignment distances. This can then be used in a modular k-nearest neighbor classifier, or k-means clusterer, to get dtw-distance knn or dtw-distance k-means, using the `dtw-python` dtw distances.

Minor changes:
* refactored the `test_has_common_interface` test which now obtains the list of methods that need to be available from the new `check_estimators._list_required_methods` method

### CODE REVIEW

**Comment by TonyBagnall** [comment]
We are in the process of converting all of the elastic distance functions to numba, you might want to wait a bit with this?

**Comment by fkiraly** [comment]
I don't think there is a conditionality? I thought we were moving to the new interface *and* numba.

**Review by Lovkush-A — CHANGES_REQUESTED** [review]
And so begins the review process. Overall structure seems to make sense, but have not checked thoroughly yet.

So far, I have thoroughly checked the helper function in utils. Some typos, some minor suggestions, and one question involving NaN indices.

To help understand what is going on, it would be good to have a notebook.

File: sktime/alignment/utils/utils_align.py
Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py:27]: clarified, is this better?

File: sktime/alignment/utils/utils_align.py
Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py:33]: I added the output

File: sktime/alignment/utils/utils_align.py
Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py:26]: fixed

File: sktime/alignment/utils/utils_align.py
Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py:41]: yes, I changed the logic. Since in all uses the original new index is used, I removed that behaviour and superfluous code.

File: sktime/alignment/utils/utils_align.py
Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py:54]: yes, that's right - fixed

File: sktime/alignment/utils/utils_align.py
Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py:95]: yes, done

File: sktime/alignment/utils/utils_align.py
Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py:77]: done

File: sktime/alignment/utils/utils_align.py
Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py:99]: introduced some changes to address this.
We also only look at the index of `Xi`. now.

**Comment by fkiraly** [comment]
I don't get what's breaking here with `black` - locally, no changes are made, and remotely, no indication of which line to change is given.

File: sktime/alignment/utils/utils_align.py
Comment (Lovkush-A) [reply sktime/alignment/utils/utils_align.py]: yup, looks good. (there might be a typo. have `data frame` in line 27)

File: sktime/alignment/utils/utils_align.py
Comment (Lovkush-A) [reply sktime/alignment/utils/utils_align.py]: looks good!

File: sktime/alignment/utils/utils_align.py
Comment (Lovkush-A) [inline sktime/alignment/utils/utils_align.py:85]:
```suggestion
    if copy:
        align = align.copy()

    if not len(X) == len(align.columns):
        raise ValueError(
```
Minor: The variable `n` was unnecessary

Comment (fkiraly) [reply sktime/alignment/utils/utils_align.py]: yes, it's no longer used anywhere else

**Review by Lovkush-A — CHANGES_REQUESTED** [review]
Looked through base class. Structure all looks good. Few aesthetic suggestions/docstring suggestions in conversations below. Couple of other comments:
* The dataframes giving the alignments seem fundamentally hard to visualise. I am not sure there is any solution to this. I think the best we can do is have a nice notebook to show examples.
* Will need unittests before merging!

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:23]:
```suggestion
        if multivariate series are passed:
```
for consistency

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:25]:
```suggestion
            if this is not set, defaults to the first variable of X[0]
```
* fixing grammar by removing `is used`
* I think we are aligning first variable in X[0] rather than first variable of X.

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:32]: * I dont think the phrase `pointwise (local) distance function` is clear / unambiguous.
* What does word local mean? From pure maths, local tends to mean that there exists a neighbourhood in which things are defined/valid. I dont think thats what is going on here.
* Giving the signature of distance function would help, e.g. `pointwise distance function with signature (float, float) -> float`

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:47]:
```suggestion
        default: first variable in X[0] as passed to fit
```
assuming my understanding from above is correct.

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:78]: need to update docstrings, to describe what dtw does and/or what this wrapper does.

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:100]:
```suggestion
        # shorthands for 1st and 2nd series
```

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:105]:
```suggestion
```
should move this reading from state with all the other variables being read from state.

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:102]: this is confusing... why change the python convention?
```suggestion
        X0 = X[0]
        X1 = X[1]
```

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:119]: instead of `vec`, it would be clearer to say what the machine type is. so either `X1_np` or `X1_list` depending on if it is numpy array or python list.

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:87]:
```suggestion
        alignment - computed alignment from dtw package
```
might be worth saying something about the (machine) type of `alignment`

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:181]:
```suggestion
        # since dtw does only pairwise alignments, this is always a 2x2 matrix
```

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:179]:
```suggestion
        distmat: a (2 x 2) np.array of floats
            [i,j]-th entry is alignment distance between X[i] and X[j] passed to fit
```

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [inline sktime/alignment/dtw_python.py:226]:
```suggestion
        window_type="None",
```

**Review by Lovkush-A — CHANGES_REQUESTED** [review]
Looked at dtw wrapper. Not really much you can do differently here. Bunch of suggestions are aesthetics/typos etc.

One major suggestion is to combine the two classes into one somehow.

* Right now, the names of the two classes do not do a good job distinguishing them and I am not sure there are good names to easily distinguish them. (When I first saw the names, I thought `AlignerDTWdist` was going to be a distance object that calculates distances using dtw)
* One possibility is that `dist_method` can either be a string or a pairwise transformer object
  * will create if loops, but also much less copying and pasting of code between the two
  * one way to make things tidier and more uniform is to create the distance matrix in both the string and transformer object case, which is then passed to dtw.

**Review by Lovkush-A — CHANGES_REQUESTED** [review]
Looked at Distance from Aligner class. This looks all fine. There seem to be two notions of symmetric going on which is confusing, and I think a minor typo to be fixed.  Also, as with other files, want to update docstrings to talk about what this particular class is doing rather than what base class is doing.

**Comment by fkiraly** [comment]
Thanks for the extensive review, @Lovkush-A! Will answer in the discussions and below.

**Comment by fkiraly** [comment]
@Lovkush-A, I've also added tests and checks now, please check the checks and tests the tests.

**Comment by Lovkush-A** [comment]
> Correctness of implementation is a concern - but not statistical properties

I do not understand the distinction you are making.

* Having correct "statistical properties" is a part of a correct implementation, no?
* Why *wouldnt* we want to check that aligners produce the 'identity alignment' when you try to align a series with itself? Is this not worth checking?

Greater issue arising from this is that we need to establish a (semi-)formal way of determining if unit-tests are adequate. Right now, it is too ad-hoc and dependent on individual reviewers and developers to decide what is enough.

**Comment by fkiraly** [comment]
@Lovkush-A, the typo is long fixed, so I would assume you approve now?

**Comment by Lovkush-A** [comment]
@fkiraly Yes, I approve! Apologies for the delay and thanks for the nudge.

I think an issue should be created to list things that arise from this PR:
* creating a notebook demo 
* having more comprehensive unittests
  * e.g. covering cases with multiple alignment
  * e.g. testing the dtw aligner's contract somehow. (note, I agree that my suggestion above - checking for identity alignment - is not fully sound. I think one option would be to create a mock object for the dtw package we wrap)

If you agree, I can create the issue.

File: sktime/alignment/dtw_python.py
Comment (Lovkush-A) [reply sktime/alignment/dtw_python.py:80]: Resolving so PR can be merged. I will add this to list of things in new issue.

**Review by Lovkush-A — APPROVED** [review]
happy with changes made in response to my various reviews.

**Comment by RNKuhns** [comment]
@fkiraly how does this fit with #1574. 

I see your PR is wrapping dtw_python. This is a popular DTW library with a variety of useful features. 

It's not clear to me why we would have functionality to implement DTW related things that seem to approach it in 2 separate ways (one wrapping an existing package and the other implementing it internally in sktime.

**Comment by TonyBagnall** [comment]
hi Ryan, I have not been involved in this. We have our own DTW and I am against introducing a dependency for dtw_python, it is pointless code duplication. Why not use the existing code in sktime?

**Comment by fkiraly** [comment]
@tonybagnall, I think your comment comes at a late stage and may conflate the concept of DTW *aligners* and the concept of *distances*.

What you are implementing are *distances* not *aligners*.
On a subtle note, you can obtain distances from aligners, but in generall not the other way round.

I think we should have DTW on the level of *aligners*, and to the best of my knowledge none of what you are currently implementing will provide this.

**Comment by TonyBagnall** [comment]
I am not in favour of a new dependency for this, or if you must have it it should be a soft dependency.

**Comment by fkiraly** [comment]
@tonybagnall, I think it's a soft dependency.
The "must" comes from this being one of the few solid implementations of time warping aligners out there.

**Comment by fkiraly** [comment]
> if you agree, I can create the issue.

yes, agreed - happy to take on some of these, sequentially.

**Comment by TonyBagnall** [comment]
perhaps make this PR draft? I wouldn't want you to sneak it in ... quoting you there, in case you have forgotten :)

**Comment by fkiraly** [comment]
@tonybagnall, it's not a draft, it's done - please treat this PR as such.
If you want changes, make a change request.

If you disagree on the grounds that "there is already DTW", I think you are mistaken on the factual basis since you are conflating DTW *aligners* with DTW *distances*, this is not about distances primarily. It would be refusing to merge a new clustering module since you are saying "there are already esteimators for time series".

**Comment by fkiraly** [comment]
> I wouldn't want you to sneak it in

PS: I'm not the person doing this, this PR is open since *AUGUST*.

On the other hand, you asked for my opinion on #1604 and then quickly merged it despite some questions not having been answered, in my opinion.

**Comment by fkiraly** [comment]
@RNKuhns, see above - this is about DTW *aligners* not DTW *distances* primarily, does that resolve your concern?

**Comment by TonyBagnall** [comment]
dismissed review on @fkiraly request

**Comment by fkiraly** [comment]
that's not what I was requesting, @TonyBagnall!
I think discussing dtw dependencies is valid; but the implicit assumption (DTW aligners vs distances) behind your comment ("we already have this") is incorrect.
If your concerns have been resolved, I'm happy, obviously, but I don't want to end this on a conflicting note.

**Comment by fkiraly** [comment]
> This is a popular DTW library with a variety of useful features.

@RNKuhns, does the above discussion alleviate your concerns?

this PR (and the `dtw-python` package) is about alignment algorithms which are different conceptually from distances (the subject of PR #1574). You can obtain distances from certain alignment algorithms (DTW), but in general not vice versa.
Further, not all alignment algorithms are DTW based, and not all distances are based on a DTW alignment. Not all alignment algorithms can be used to obtain a distance either. These are conceptually related, but conceptually different.

Conceptually, a distance (between time series) models a mathematical object d: series x series -> R
==> this PR
an alignment algorithm models a mathematical object a: series x series -> alignment
==> PR #1574

As said: related, but different concepts.
Like supervised regression and forecasting are related but different; and tied together by reduction/composition.

The reason for common conceptual conflation on "DTW" is because of various reduction relations in the space. People say DTW, and sometimes mean the *aligner*, sometimes the *distance*, sometimes a *classifier* or *clusterer* that uses the *distances*.

**Comment by fkiraly** [comment]
perhaps a clearer name for this PR would be "base class for alignment algorithms (aligners) and distances reduced to aligners", but I chose the name DTW above the names for abstract algorithm types, since I thought that's something people can (apparently) better relate to...

**Comment by fkiraly** [comment]
@RNKuhns, @TonyBagnall didn't attend today's dev meeting where this PR was on the agenda.

I'm interpreting this discussion as follows:
* @Lovkush-A is fine with this
* @RNKuhns had a question which he deferred to @TonyBagnall 
* @TonyBagnall has had a change request, but that got resolved above in discussion, and he is not opposed to this being merged
* no one else has looked at this properly, with the PR open since Aug which is ample time

Hence it fulfills the criteria for acceptance, and I'm merging this now.

**Pull request merged**
